modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/PartitionMetaManager.java

```diff
@@ -0,0 +1,168 @@
+import static org.apache.ignite.internal.util.GridUnsafe.freeBuffer;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.ignite.internal.pagememory.io.PageIoRegistry;
+import org.apache.ignite.internal.pagememory.persistence.PartitionMeta.PartitionMetaSnapshot;
+import org.apache.ignite.internal.pagememory.persistence.io.PartitionMetaIo;
+import org.apache.ignite.internal.pagememory.persistence.store.FilePageStore;
+import org.apache.ignite.lang.IgniteInternalCheckedException;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Partition meta information manager.
+ */
+// TODO: IGNITE-17132 Do not forget about deleting the partition meta information
+public class PartitionMetaManager {
+    private final Map<GroupPartitionId, PartitionMeta> metas = new ConcurrentHashMap<>();
+
+    private final PageIoRegistry ioRegistry;
+
+    private final int pageSize;
+
+    /**
+     * Constructor.
+     *
+     * @param ioRegistry Page IO Registry.
+     * @param pageSize Page size in bytes.
+     */
+    public PartitionMetaManager(PageIoRegistry ioRegistry, int pageSize) {
+        this.ioRegistry = ioRegistry;
+        this.pageSize = pageSize;
+    }
+
+    /**
+     * Returns the partition's meta information.
+     *
+     * @param groupPartitionId Partition of the group.
+     */
+    public @Nullable PartitionMeta getMeta(GroupPartitionId groupPartitionId) {
+        return metas.get(groupPartitionId);
+    }
+
+    /**
+     * Adds partition meta information.
+     *
+     * @param groupPartitionId Partition of the group.
+     * @param partitionMeta Partition meta information.
+     */
+    public void addMeta(GroupPartitionId groupPartitionId, PartitionMeta partitionMeta) {
+        metas.put(groupPartitionId, partitionMeta);
+    }
+
+    /**
+     * Reads the partition {@link PartitionMeta meta} from the partition file or creates a new one.
+     *
+     * <p>If it creates a new one, it writes the meta to the file.
+     *
+     * @param checkpointId Checkpoint ID.
+     * @param groupPartitionId Partition of the group.
+     * @param pageMemory Page memory.
+     * @param filePageStore Partition file page store.
+     */
+    public PartitionMeta readOrCreateMeta(
+            @Nullable UUID checkpointId,
+            GroupPartitionId groupPartitionId,
+            PersistentPageMemory pageMemory,
+            FilePageStore filePageStore
+    ) throws IgniteInternalCheckedException {
+        ByteBuffer buffer = allocateBuffer(pageSize);
+
+        long bufferAddr = bufferAddress(buffer);
+
+        long partitionMetaPageId = pageMemory.partitionMetaPageId(groupPartitionId.getGroupId(), groupPartitionId.getPartitionId());
+
+        try {
+            if (filePageStore.size() > filePageStore.headerSize()) {
+                // Reads the partition meta.
+                if (filePageStore.pages() == 0) {
```

Comment:
This is a tricky question. I think that a specific method for reading pages is required, that will check file size instead of logical pages count. So it's not 1 or 3, it's more like 2.1.1

Reply:
Could this affect performance if we ask the operating system for the file size when reading pages?

Reply:
I don't know actually. I wouldn't risk it for every page. Especially considering that in the future file size will not match the logical pages count if you had failed checkpoint.
BTW, I think we should double-check fault tolerance of creating a new partition file. It might not work if we failed to write meta page I think. Also, please check that you didn't forget to fsync file after you write meta page into it.

Reply:
I don't understand what you are suggesting.